unattended-upgrade

```diff
@@ -1256,24 +1269,88 @@ def get_auto_removable(cache):
                 if pkg.is_auto_removable])
 
 
-def do_auto_remove(cache, auto_removable, logfile_dpkg,
+def only_kernel_pkgs(auto_removable):
+    # type: (AbstractSet[str]) -> AbstractSet[str]
+    running_kernel_version = subprocess.check_output(
+        ["uname", "-r"], universal_newlines=True).rstrip()
+    versioned_kernel_pkgs = apt_pkg.config.value_list(
+        "APT::VersionedKernelPackages")
+
+    if versioned_kernel_pkgs:
+        kernel_pkgs_regexp = "(" + "|".join(
+            ["^" + p for p in versioned_kernel_pkgs]) + ")"
+        logging.debug("Using %s regexp to find kernel packages",
+                      kernel_pkgs_regexp)
+
+        return set([pkg for pkg in auto_removable
+                    if (re.match(kernel_pkgs_regexp, pkg) and
+                        not re.match(running_kernel_version, pkg))])
+    else:
+        logging.debug("APT::VersionedKernelPackages is not set")
+        return set()
+
+
+def do_auto_remove(cache, auto_removable, logfile_dpkg, minimal_steps,
                    verbose=False, dry_run=False):
-    # type: (apt.Cache, Iterable[str], str, bool, bool) -> bool
+    # type: (apt.Cache, AbstractSet[str], str, bool, bool, bool) -> bool
     if not auto_removable:
         return True
 
-    for pkgname in auto_removable:
-        if not dry_run:
-            logging.debug("marking %s for remove" % pkgname)
+    if minimal_steps:
+        for pkgname in auto_removable:
+            if SIGNAL_STOP_REQUEST:
+                logging.warning("SIGNAL received, stopping")
+                return False
+            logging.debug("marking %s for removal" % pkgname)
+            cache[pkgname].mark_delete()
+            changes = set([p.name for p in cache.get_changes()])
```

Comment:
This looks quite similar to the code below - maybe a small helper can be extracted here?

Reply:
It is similar, but does not seem like a good candidate for abstracting it out and adding a new redirection, but I'm not absolutely sure I got your idea right.

Reply:
I was thinking about something like: http://paste.ubuntu.com/p/3D4Q7rSX6y/ - I understand the generated message is slightly different so maybe its overkill. Or a slightly modified version with a different error message http://paste.ubuntu.com/p/3WsGxyjMXw/ - please let me know if I overlooked something (maybe not enough tea yet ;)

Reply:
Ah, OK, I see now. Feel free to add a commit on top of this.